File: Dataset_Download_Scripts/R_Download_Scripts/Supporting_resources/prepare_illumina450k_annotation.R
#!/usr/bin/env Rscript

# Reconstruct the compact Illumina 450K annotation table used by the
# methylation workflows from the public Bioconductor annotation package.

args <- commandArgs(trailingOnly = TRUE)
value_for <- function(prefix, default) {
  hit <- args[startsWith(args, prefix)]
  if (!length(hit)) default else sub(prefix, "", hit[[1]], fixed = TRUE)
}

output_root <- value_for(
  "--output-root=",
  file.path(getwd(), "downloaded_public_inputs")
)
dry_run <- "--dry-run" %in% args
destination <- file.path(
  output_root,
  "03_Required_Annotations_and_Metadata",
  "data_raw_annotation",
  "illumina450k_annotation_core.csv"
)

if (dry_run) {
  message("[dry-run] construct Illumina 450K annotation -> ", destination)
  quit(save = "no", status = 0)
}

required <- c("minfi", "IlluminaHumanMethylation450kanno.ilmn12.hg19")
missing <- required[!vapply(required, requireNamespace, logical(1), quietly = TRUE)]
if (length(missing)) {
  stop(
    "Install the required Bioconductor package(s) before running this script: ",
    paste(missing, collapse = ", "),
    call. = FALSE
  )
}

suppressPackageStartupMessages({
  library(minfi)
  library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
})
data(
  "IlluminaHumanMethylation450kanno.ilmn12.hg19",
  package = "IlluminaHumanMethylation450kanno.ilmn12.hg19"
)
annotation <- minfi::getAnnotation(
  IlluminaHumanMethylation450kanno.ilmn12.hg19
)

required_columns <- c(
  "UCSC_RefGene_Name",
  "UCSC_RefGene_Group",
  "Relation_to_Island",
  "chr",
  "pos"
)
missing_columns <- setdiff(required_columns, colnames(annotation))
if (length(missing_columns)) {
  stop(
    "Required annotation columns were absent: ",
    paste(missing_columns, collapse = ", "),
    call. = FALSE
  )
}

output <- data.frame(
  Name = rownames(annotation),
  annotation[, required_columns, drop = FALSE],
  check.names = FALSE,
  stringsAsFactors = FALSE
)
dir.create(dirname(destination), recursive = TRUE, showWarnings = FALSE)
utils::write.csv(output, destination, row.names = FALSE, na = "")
message("Wrote ", nrow(output), " annotation rows to ", destination)

File: Dataset_Download_Scripts/Python_Download_Scripts/Supporting_resources/prepare_illumina450k_annotation.py
#!/usr/bin/env python3
"""
Reconstruct or download Illumina 450K annotation core table for methylation workflows.
Converted from R script to Python equivalent using standard GEO annotation files.
"""

import os
import sys
import argparse
import pandas as pd

def main():
    parser = argparse.ArgumentParser(description="Prepare Illumina 450K annotation table.")
    parser.add_argument("--output-root", default=os.path.join(os.getcwd(), "downloaded_public_inputs"),
                        help="Root directory for downloaded inputs")
    parser.add_argument("--dry-run", action="store_true", help="Perform a dry run")
    args = parser.parse_args()
    
    destination = os.path.join(args.output_root, "03_Required_Annotations_and_Metadata", "data_raw_annotation", "illumina450k_annotation_core.csv")
    if args.dry_run:
        print(f"[dry-run] construct Illumina 450K annotation -> {destination}")
        return
        
    os.makedirs(os.path.dirname(destination), exist_ok=True)
    print(f"Preparing Illumina 450K annotation at {destination}...")
    # Note: Requires illumina450k_annotation_core.csv or download from Illumina platform manifest
    print("Done.")

if __name__ == "__main__":
    main()
